File: coding-test/easy/Majority Element.py
"""
Approach
To solve the problem in linear time and O(1) space, we can use Boyer-Moore Voting Algorithm. 
The algorithm works by maintaining a count variable and a candidate variable. 
We iterate through the array, for each element, we check if the count is 0, if it is, we update the candidate variable to the current element. 
We then increment the count variable if the current element is equal to the candidate, otherwise we decrement the count variable. 
At the end of the iteration, the candidate variable will hold the majority element.

Complexity
The time complexity of this algorithm is O(n), where n is the length of the array.
The space complexity is O(1), as we only need to maintain two variables.
"""


class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        # initialize the count and candidate variables
        count = 0
        candidate = None

        # iterate through the array
        for num in nums:
            # if count is 0, update the candidate variable
            if count == 0:
                candidate = num
            # increment or decrement count based on whether the current element is the candidate
            count += 1 if num == candidate else -1

        # return the candidate variable
        return candidate


# Follow-up: Could you solve the problem in linear time and in O(1) space?


# time O(n), space O(n)
class Solution2:
    def majorityElement(self, nums: list[int]) -> int:
        counter = {}

        for num in nums:
            if num in counter:
                counter[num] += 1
            else:
                counter[num] = 1

        max_count = -1
        ans = -1

        for key, val in counter.items():
            if val > max_count:
                max_count = val
                ans = key

        return ans

File: coding-test/TS_CS/ConsoleApp1/Majority Element.ts
const majorityElement = (nums: number[]) => {
    let map: { [key: number]: number } = {}
    const majorityCount = Math.floor(nums.length / 2)

    for (const num of nums) {
        // if (map[num]) {
        //     map[num]++
        //     if (map[num] > majorityCount) {
        //         return num
        //     }
        // } else {
        //     map[num] = 1
        // }

        map[num] = (map[num] ?? 0) + 1
        if (map[num] > majorityCount) {
            return num
        }
    }
}
// 시간 복잡도: O(n) 배열을 한 번 순회하며 map을 업데이트.
// 공간 복잡도: O(n) 숫자와 그 빈도를 저장하기 위한 map의 공간.

console.log(majorityElement([3, 2, 3])); // 3
console.log(majorityElement([2, 2, 1, 1, 1, 2, 2])); // 2

function majorityElementBoyerMoore(nums: number[]): number {
    let candidate = nums[0];
    let count = 0;

    for (const num of nums) {
        if (count === 0) {
            candidate = num; // 새로운 후보 선정
        }
        count += (num === candidate) ? 1 : -1; // 후보와 같으면 증가, 다르면 감소
    }

    return candidate;
}

// 보이어-무어 알고리즘:

// 효율적이고 O(1) 공간 복잡도.
// 대규모 데이터에 적합.